packages/components/tooltip/src/tooltip.tsx

```diff
@@ -1,49 +1,153 @@
+   */
+  styles: Record<string, SerializedStyles>;
+  /**
+   * Determines if the tooltip should not appear.
+   */
+  off: boolean;
+  /**
+   * An identifier for the tooltip, used for `aria-describedby`.
+   */
+  id?: string;
+  /**
+   * A callback function, called when the tooltip is closing.
+   */
+  onClose?: (e?: ChangeEvent | FocusEvent) => void;
+  /**
+   * A callback function, called when the tooltip is opening.
+   */
+  onOpen?: (e?: ChangeEvent | FocusEvent) => void;
+  isOpen?: boolean;
+  /**
+   * How the tooltip is positioned relative to the child element.
+   */
+  placement?:
+    | 'top'
+    | 'top-start'
+    | 'top-end'
+    | 'right'
+    | 'right-start'
+    | 'right-end'
+    | 'bottom'
+    | 'bottom-start'
+    | 'bottom-end'
+    | 'left'
+    | 'left-start'
+    | 'left-end';
+  /**
+   * The message to show in the tooltip.
+   */
+  title: string;
+  /**
+   * Provides a way to fine-tune an appearance of underlying Popper tooltip element. For more information, please check [Popper.js documentation](https://popper.js.org/popper-documentation.html#modifiers).
+   */
+  modifiers?: Modifiers;
+  /**
+   * Customize the appearance of certain elements of the tooltip.
+   */
+  components?: TComponents;
+  /**
+   * Horizontal size limit of the tooltip.
+   */
+  horizontalConstraint?:
+    | 1
+    | 2
+    | 3
+    | 4
+    | 5
+    | 6
+    | 7
+    | 8
+    | 9
+    | 10
+    | 11
+    | 12
+    | 13
+    | 14
+    | 15
+    | 16
+    | 'scale'
+    | 'auto';
+};
+
+const TooltipWrapper = (props: Pick<TTooltipProps, 'children'>) => (
   <React.Fragment>{props.children}</React.Fragment>
 );
 TooltipWrapper.displayName = 'TooltipWrapperComponent';
 TooltipWrapper.propTypes = {
   children: PropTypes.node.isRequired,
 };
 
-const Tooltip = (props) => {
-  const leaveTimer = React.useRef();
-  const childrenRef = React.useRef();
+const tooltipDefaultProps: Pick<
+  TTooltipProps,
+  'closeAfter' | 'horizontalConstraint' | 'off' | 'placement'
+> = {
+  closeAfter: 0,
+  horizontalConstraint: 'scale',
+  off: false,
+  placement: 'top',
+};
 
-  warning(
-    props.components.BodyComponent
-      ? isValidElementType(props.components.BodyComponent)
-      : true,
-    `ui-kit/Tooltip: the prop 'components.BodyComponent' is not a valid React element.`
-  );
-  warning(
-    props.components.TooltipWrapperComponent
-      ? isValidElementType(props.components.TooltipWrapperComponent)
-      : true,
-    `ui-kit/Tooltip: the prop 'components.TooltipWrapperComponent' is not a valid React element.`
-  );
-  warning(
-    props.components.WrapperComponent
-      ? isValidElementType(props.components.WrapperComponent)
-      : true,
-    `ui-kit/Tooltip: the prop 'components.WrapperComponent' is not a valid React element.`
-  );
+const Tooltip = (props: TTooltipProps) => {
+  const leaveTimer = React.useRef<ReturnType<typeof setTimeout>>();
+  const childrenRef = React.useRef<ReturnType<typeof setTimeout>>();
```

Comment:
I get the usage of `typeof setTimeout` in the previous line but is it correct also in this line?

Reply:
yes.